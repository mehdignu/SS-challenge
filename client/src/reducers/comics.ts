import { Comic, FetchComicsAction } from '../actions';
import { ActionTypes } from '../actions/types';

export const comicsReducer = (state: Comic[] = [], action: FetchComicsAction) => {
  switch (action.type) {
    case ActionTypes.fetchComics:
      return action.payload;
    default:
      return state;
  }
};
