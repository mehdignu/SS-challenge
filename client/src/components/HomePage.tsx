import React from 'react';
import { connect } from 'react-redux';
import { Comic, fetchComics } from '../actions';
import { StoreState } from '../reducers';
import { Link } from 'react-router-dom';
import moment from 'moment';

interface LykonProps {
  comics: Comic[];
  fetchComics(): any;
}

class HomePage extends React.Component<LykonProps> {
  renderComics(): JSX.Element[] {
    return this.props.comics
      .sort((d1, d2) => {
        const a = new Date(
          moment(d1.publishDate, 'DD.MM.YYYY').format('YYYY-MM-DD')
        );
        const b = new Date(
          moment(d2.publishDate, 'DD.MM.YYYY').format('YYYY-MM-DD')
        );
        return a < b ? 1 : -1;
      })
      .map((comic: Comic) => {
        return (
          <div key={comic.id}>
            {comic.title}
            <br />
            <Link to={`comics/${comic.id}`}>Show Post</Link>
            <br />
            {comic.publishDate}
          </div>
        );
      });
  }

  render() {
    return <div>{this.renderComics()}</div>;
  }
}

const mapStateToProps = ({ comics }: StoreState): { comics: Comic[] } => {
  return { comics };
};

export default connect(mapStateToProps, { fetchComics })(HomePage);
