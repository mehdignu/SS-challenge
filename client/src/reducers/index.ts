import { combineReducers } from 'redux';
import { Comic } from '../actions';
import { comicsReducer } from './comics';

export interface StoreState {
  comics: Comic[];
}

export const reducers = combineReducers({
  comics: comicsReducer
});
