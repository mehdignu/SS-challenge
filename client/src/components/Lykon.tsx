import React from 'react';
import { connect } from 'react-redux';
import { Comic, fetchComics } from '../actions';
import { StoreState } from '../reducers';
import {BrowserRouter, Switch, Route} from 'react-router-dom';
import HomePage from './HomePage';
import ComicDetails from './ComicDetails';

interface LykonProps {
  comics: Comic[];
  fetchComics(): any;
}

class Lykon extends React.Component<LykonProps> {

  componentDidMount() {
    this.props.fetchComics();
  }

  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path='/' component={HomePage}  />
          <Route exact path='/comics/:comicId' component={ComicDetails} />
        </Switch>
      </BrowserRouter>
    );
  }
}

const mapStateToProps = ({ comics }: StoreState): { comics: Comic[] } => {
  return { comics };
};

export const App = connect(mapStateToProps, { fetchComics })(Lykon);
