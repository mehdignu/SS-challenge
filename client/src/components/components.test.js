import expect from 'expect';
import { shallow, configure, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import HomePage from './HomePage';
import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { App } from './Lykon';
import { reducers } from '../reducers';

const store = createStore(reducers, applyMiddleware(thunk));

configure({ adapter: new Adapter() });

it('should expect to render the App component', () => {
  const wrapper = mount(
    <Provider store={store}>
      <App />
    </Provider>
  );
  expect(wrapper.debug()).toMatchSnapshot();
});

it('should expect to render the HomePage component', () => {
  expect(
    shallow(
      <Provider store={store}>
        <HomePage />
      </Provider>
    ).length
  ).toEqual(1);
});
