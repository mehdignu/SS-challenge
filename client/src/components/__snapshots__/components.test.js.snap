// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should expect to render the App component 1`] = `
"<Provider store={{...}}>
  <Connect(Lykon)>
    <Lykon comics={{...}} fetchComics={[Function (anonymous)]}>
      <BrowserRouter>
        <Router history={{...}}>
          <Switch>
            <Route exact={true} path=\\"/\\" component={{...}} location={{...}} computedMatch={{...}}>
              <Connect(HomePage) history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
                <HomePage history={{...}} location={{...}} match={{...}} staticContext={[undefined]} comics={{...}} fetchComics={[Function (anonymous)]}>
                  <div />
                </HomePage>
              </Connect(HomePage)>
            </Route>
          </Switch>
        </Router>
      </BrowserRouter>
    </Lykon>
  </Connect(Lykon)>
</Provider>"
`;
